File: CS330/1_operator_overloading/division.cpp
#include "division.h"
#include "project.h"
#include "split.h"

using namespace std;


Division::Division() {}


void Division::addProject(Project b)
{
  activeProjects.push_back (b);
}


int Division::numberOfProjects() const
{
  return activeProjects.size();
}


  
istream& operator>> (istream& in, Division& p)
{
  string line;
  if (in)
    {
      getline(in, line);
    }

  string fields[4];
  int fieldCount = split(line, "\t", fields, 4);

  p = Division();
  p.putName (fields[0]);
  p.putRoom (fields[1]);
  p.putBuilding (fields[2]);
  p.putDivisionCode (fields[3]);
}
      

std::ostream& operator<< (std::ostream& out, const Division& div)
{
  out << div.getName() << "\t"
      << div.getDivisionCode() << "\t"
      << div.getBuilding() << ":" << div.getRoom()
      << "\n";
  if (div.numberOfProjects() > 0)
    {
      int i = 0;
      for (Division::const_iterator p = div.begin();
	   p != div.end(); ++p)
	{
	  ++i;
	  out << i << "/" << div.numberOfProjects() << ": ";
	  out << *p << "\n";
	}
    }
  else
    out << "(no projects)\n";
  return out;
}






Division::iterator Division::begin()
{
  return activeProjects.begin();
}

Division::const_iterator Division::begin() const
{
  return activeProjects.begin();
}

Division::iterator Division::end()
{
  return activeProjects.end();
}

Division::const_iterator Division::end() const
{
  return activeProjects.end();
}

File: CS330/4_test_project_in_java/Division.java
import java.io.BufferedReader;
import java.util.ArrayList;
import java.util.Iterator;

/**
 * Describes a division of the company.
 */

public class Division implements Comparable<Division>{
	private String theName;
	private String theRoom;
	private String theBuilding;
	private String theDivisionCode;

	private ArrayList<Project> activeProjects;

	public Division() {
		theName = "";
		theRoom = "";
		theBuilding = "";
		theDivisionCode = "";
		activeProjects = new ArrayList<Project>();
	}

//	Attributes
	public String getName() {return theName;}
	public void putName(String name) {theName = name;}

	public String getRoom() {return theRoom;}
	public void putRoom(String room) {theRoom = room;}

	public String getBuilding() {return theBuilding;}
	public void putBuilding(String building) {theBuilding = building;}

	public String getDivisionCode() {return theDivisionCode;}
	public void putDivisionCode(String divisionCode) {theDivisionCode = divisionCode;}

	public void addProject(Project proj)
	{
		assert (proj.getClass().getName().contains("Project"));
		activeProjects.add(proj);
	}

	public int numberOfProjects()
	{
		return activeProjects.size();
	}

	public static Division read (BufferedReader in) throws java.io.IOException
	{
		String line = in.readLine();
		if (line == null)
			return null;
		String fields[] = line.split("\t");
		if (fields.length != 4)
			throw new java.io.IOException ("Improper input format for Division");
		  Division d = new Division();
		  d.putName (fields[0]);
		  d.putRoom (fields[1]);
		  d.putBuilding (fields[2]);
		  d.putDivisionCode (fields[3]);
		  return d;
	}

	public Iterator<Project> iterator()
	{
		return activeProjects.iterator();
	}

	public int compareTo(Division d) {
		int result = theName.compareTo (d.theName);
		if (result == 0)
			result = theDivisionCode.compareTo(d.theDivisionCode);
		if (result == 0)
			result = theBuilding.compareTo(d.theBuilding);
		if (result == 0)
			result = theRoom.compareTo(d.theRoom);
		return result;
	}

	public boolean equals(Object d) {
		Division div = (Division)d;
		return (compareTo(div) == 0);
	}
	
	public int hashCode()
	{
		int result = theName.hashCode();
		result = 3 * result + theDivisionCode.hashCode();
		result = 3 * result + theBuilding.hashCode();
		result = 3 * result + theRoom.hashCode();
		return result;
	}
	
	public String toString()
	{
		StringBuffer buf = new StringBuffer();
		buf.append (theName);
		buf.append (", ");
		buf.append (theDivisionCode);
		buf.append (", ");
		buf.append (theBuilding);
		buf.append (", ");
		buf.append (theRoom);
		buf.append (": ");
		Iterator<Project> it = activeProjects.iterator();
		while (it.hasNext()) {
			Project p = it.next();
			buf.append (p.getTitle());
			buf.append(", ");
		}
		return buf.toString();
	}

}



